dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/writer/JSONMetricWriter30.java

```diff
@@ -0,0 +1,174 @@
+                    break;
+                case MetricUnits.DAYS:
+                    conversionFactor = 24 * 60 * 60 * 1000000000L;
+                    break;
+            }
+
+            if (tagsMap.size() != 0) {
+                for (Entry<String, String> tagsMapEntrySet : tagsMap.entrySet()) {
+                    String tagValue = tagsMapEntrySet.getValue();
+
+                    /*
+                     * As per MicroProfile metrics spec
+                     * Need to convert semicolons found in the value of tags
+                     * to underscores
+                     */
+                    if (tagValue.contains(";")) {
+                        tagValue = tagValue.replaceAll(";", "_");
+                    }
+
+                    tags += ";" + tagsMapEntrySet.getKey() + "=" + tagValue;
+                }
+                metricNameWithTags = metricName + tags;
+            }
+            Metric metric = entry.getValue();
+
+            if (Counter.class.isInstance(metric)) {
+                jsonObject.put(metricNameWithTags, ((Counter) metric).getCount());
+            } else if (ConcurrentGauge.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMap(Util30.getConcurrentGaugeNumbers((ConcurrentGauge) metric, tags), metricName, jsonObject));
+            } else if (Gauge.class.isInstance(metric)) {
+                try {
+                    jsonObject.put(metricNameWithTags, ((Gauge) metric).getValue());
+                } catch (IllegalStateException e) {
+                    // The forwarding gauge is likely unloaded. A warning has already been emitted
+                }
+            } else if (Timer.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMap(Util30.getTimerNumbers((Timer) metric, tags, conversionFactor), metricName, jsonObject));
+            } else if (SimpleTimer.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMapAsObject(Util30.getSimpleTimerNumbersAsObjects((SimpleTimer) metric, tags, conversionFactor), metricName, jsonObject));
+            } else if (Histogram.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMap(Util30.getHistogramNumbers((Histogram) metric, tags), metricName, jsonObject));
+            } else if (Meter.class.isInstance(metric)) {
+                jsonObject.put(metricName, getJsonFromMap(Util30.getMeterNumbers((Meter) metric, tags), metricName, jsonObject));
+            } else {
+                Tr.event(tc, "Metric type '" + metric.getClass() + " for " + metricName + " is invalid.");
+            }
+        }
+        return jsonObject;
+    }
+
+    /*
+     * This method compared to getJsonFromMap returns a map with a value as an Object.. This is because
+     * the value can be a NaN object.
+     */
+    protected JSONObject getJsonFromMapAsObject(Map<String, Object> metricMap, String metricName, JSONObject parentJSONObject) {
```

Comment:
Needed a new method that returns a map with Object values. This is due to the possibility of a null value being returned.